### PR TITLE
Refactor MonteCarlo to be a StatefulInterpretation

### PR DESCRIPTION
**Pull request by fritzo**
Pair coded with @eb8680 

@eb8680 I have adapted `StatefuleInterpretation` from https://github.com/pyro-ppl/funsor/tree/adam-interpreter, making it a `namedtuple` as you had originally suggested, and turning on fallback logic by default as you suggested.

This PR:
- [x] Adds a `StatefulInterpretation` base class for interpretations that need to pass instance-specific data such as num_samples and rng_key.
- [x] Adds an `InterpreterStack` and makes the `MonteCarlo` interpretation fall back not to `eager` but the previous interpretation when it was installed.
- [x] Refactors `monte_carlo` to a stateful `MonteCarlo`.
- [x] Fixes the JAX implementation of `MonteCarlo` by tracking `rng_key`
  ~~@fehiepsi can you help me out with this?~~

### CODE REVIEW

File: funsor/interpreter.py [hunk 1]
```diff
@@ -324,13 +333,50 @@ def dispatched_interpretation(fn):
...
+
+        with interpretation(MyInterpretation(my_param=0.1)):
+            ...
+    """
+    def __init_subclass__(cls):
+        cls.registry = KeyedRegistry(default=lambda *args: None)
+        cls.dispatch = cls.registry.dispatch
+
+    def __call__(self, cls, *args):
+        return self.dispatch(cls, *args)(self, *args)
```
Comment (fritzo) [inline funsor/interpreter.py:360]: This default implementation of `__call__()` relies on `InterpreterStack` to fall back to the previous interpretation, which is by assumption complete.

File: funsor/montecarlo.py [hunk 2]
```diff
@@ -2,57 +2,34 @@
...
-        monte_carlo.sample_inputs = old
-
-
-@monte_carlo.register(Integrate, Funsor, Funsor, frozenset)
-def monte_carlo_integrate(log_measure, integrand, reduced_vars):
-    # FIXME: how to pass rng_key to here?
-    sample = log_measure.sample(reduced_vars, monte_carlo.sample_inputs)
+@MonteCarlo.register(Integrate, Funsor, Funsor, frozenset)
+def monte_carlo_integrate(state, log_measure, integrand, reduced_vars):
+    # FIXME use state.rng_key to here
```
Comment (fritzo) [inline funsor/montecarlo.py:25]: @fehiepsi can you help me out with the `rng_key` logic here? I have removed the `xfail` from your test case in hopes that we can fix the issue in this PR.

File: funsor/montecarlo.py [hunk 3]
```diff
@@ -2,57 +2,34 @@
...
-    return result
-
+    expressions. This falls back to the previous interpreter in other cases.
 
-# This is a globally configurable parameter to draw multiple samples.
-monte_carlo.sample_inputs = OrderedDict()
+    :param rng_key:
+    """
+    def __init__(self, *, rng_key=None, **sample_inputs):
+        self.rng_key = rng_key
```
Comment (fritzo) [inline funsor/montecarlo.py:19]: @fehiepsi I'm not sure whether `rng_key` should be threaded through the constructor, or whether it should instead be read-and-written in pattern handlers like `monte_carlo_integrate()` below.

File: funsor/montecarlo.py (hunk 2 above)
Comment (fehiepsi) [reply funsor/montecarlo.py:25]: Do we only use `monte_carlo_integrate` one? If so, I think you can simply add `rng_key=state.rng_key` keyword:
```
sample = log_measure.sample(reduced_vars, state.sample_inputs, rng_key=state.rng_key)
```

But if this code is run multiple time, then we need to split the key:
```python
from funsor.util import get_backend
...
    if state.rng_key is not None and get_backend() == "jax":
        import jax

        rng_key, state.rng_key = jax.random.split(state.rng_key)
    sample = log_measure.sample(reduced_vars, state.sample_inputs, rng_key=rng_key)
```

Comment (fehiepsi) [reply funsor/montecarlo.py:25]: @fritzo Could you help me replace `if rng_key is not None` at [this line](https://github.com/pyro-ppl/funsor/blob/master/funsor/cnf.py#L89) and 10 lines below by `if rng_key is not None and get_backend() == "jax"`? I feel doing so is safer.

Comment (fritzo) [reply funsor/montecarlo.py]: Thanks for the help, jax tests now pass :tada:

File: funsor/interpreter.py [hunk 4]
```diff
@@ -324,13 +333,54 @@ def dispatched_interpretation(fn):
     return fn
 
 
+class StatefulInterpretationMeta(type):
```
Comment (eb8680) [inline funsor/interpreter.py:336]: I guess this is required for Python 3.5?  Once we drop Python 3.5 support, we should remove this and other metaclasses where possible in favor of `__init_subclass__`.

**Review by eb8680 — APPROVED** [review]
LGTM, `MonteCarlo` is much cleaner.

File: funsor/interpreter.py (hunk 4 above)
Comment (fritzo) [reply funsor/interpreter.py:336]: Agreed, I was almost tempted to drop support for Python 3.5 in this PR, but that seemed like overkill.

**Comment by fritzo** [comment]
@eb8680 Thanks for reviewing!
@fehiepsi Thanks for helping!

**Pull request merged**